### DISCUSSION TITLE
Allow Depends in APIRouter to Share Dependency Object Across All Route Functions

### DISCUSSION THREAD

**Discussion by CuriousGu** [fastapi/fastapi#12549] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
### How it currently works ###

from fastapi import APIRouter, Depends
from my_path import my_func

router = APIRouter()

@router.get("/get")
def get(my_var: any = Depends(my_func)):
    pass

### What I'm expecting ###

from fastapi import APIRouter, Depends
from my_path import my_func

router = APIRouter("my_route", dependencies={my_obj: Depends(my_func)})

@router.get("\get")
def get():
    return my_obj
```

### Description

I would like to suggest a feature where Depends could be used in the APIRouter declaration, allowing the returned object to be accessible within all route functions in that router.

Currently, while it’s possible to use Depends in the router, it doesn’t enable the object to be directly available inside each route function without redeclaring the dependency. This enhancement could simplify development for applications that need a shared context or instance across all functions in a router, such as a database session or a common validation that returns user's information.

### Operating System

Linux

### FastAPI Version

0.115.3

### Pydantic Version

2.9.2

### Python Version

3.11.4

**Comment by YuriiMotov** [comment, marked as answer]
FastAPI is focusing on giving us the best editor support (type annotations, auto-completion).
Do you know the way how to implement your idea considering this constraint?

Actually, you can achieve what you want (to pass parameters into endpoint function using router-dependency) in FastAPI (although without best editor support):
```
from fastapi import APIRouter, Depends, FastAPI, Request


# Define dependencies that not only return values, but also store them in request.state
async def get_session_1(request: Request):
    session = "session #1"
    request.state.session_1 = session  # add session_1 to request.state
    yield session


async def get_session_2(request: Request):
    session = "session #2"
    request.state.session_2 = session  # add session_2 to request.state
    yield session


# Include dependencies into router's dependencies
router = APIRouter(
    prefix="/my_route", dependencies=[Depends(get_session_1), Depends(get_session_2)]
)


# Access dependencies' values using request.state
@router.get("/")
def get(request: Request):
    print(request.state.session_1)
    print(request.state.session_2)


app = FastAPI()
app.include_router(router=router)

```

**Comment by CuriousGu** [comment]
I'm gonna try. Could you open an issue?

**Reply by YuriiMotov** [reply]
I think you don't need to have an open issue to try to implement what you want.
As you can see, repository owner tends to use `issues` only for approved bugs and accepted features.

You can just implement what you want and open a PR. Repository owner will decide whether it worth being included in FastAPI or not.
